File: backend/app/Models/Review.php
<?php

namespace App\Models;

// use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Review extends Model
{
    // use HasFactory;

    protected $table = 'reviews';
    protected $primaryKey = 'id';
    public $incrementing = true;
    protected $keyType = 'int';

    protected $fillable = [
        'user_id',
        'room_id',
        'booking_detail_id',
        'rating',
        'description',
        'image',
    ];

    protected $casts = [
    'rating' => 'integer',
    'description' => 'string',
    'image' => 'string',
    'created_at' => 'datetime:Y-m-d H:i:s',
    'updated_at' => 'datetime:Y-m-d H:i:s',
];

    protected $appends = ['image_url'];

    public function getImageUrlAttribute()
    {
        return $this->image
            ? asset('storage/' . $this->image)
            : null;
    }

    public function user()
    {
        return $this->belongsTo(User::class, 'user_id');
    }

    public function room()
    {
        return $this->belongsTo(Room::class, 'room_id');
    }

    public function bookingDetail()
    {
        return $this->belongsTo(BookingDetail::class, 'booking_detail_id');
    }
}

File: frontend/lib/models/review.dart
import 'package:frontend/models/review_user.dart';

class Review {
  final int id;
  final int userId;
  final int roomId;
  final int bookingDetailId;
  final int rating;
  final String description;
  final String? image;
  final String? createdAt;
  final ReviewUser? user;

  Review({
    required this.id,
    required this.userId,
    required this.roomId,
    required this.bookingDetailId,
    required this.rating,
    required this.description,
    this.image,
    this.createdAt,
    this.user,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      roomId: json['room_id'] ?? 0,
      bookingDetailId: json['booking_detail_id'] ?? 0,
      rating: json['rating'] ?? 0,
      description: json['description'] ?? '',
      image: json['image'],
      createdAt: json['created_at'],
      user: json['user'] != null ? ReviewUser.fromJson(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'room_id': roomId,
      'booking_detail_id': bookingDetailId,
      'rating': rating,
      'description': description,
      'image': image,
      'created_at': createdAt,
    };
  }
}
